game-core/src/main/java/games/strategy/triplea/ai/pro/ProCombatMoveAi.java

```diff
@@ -375,14 +367,25 @@ private void determineTerritoriesToAttack(final List<ProTerritory> prioritizedTe
   }
 
   private void determineTerritoriesThatCanBeHeld(final List<ProTerritory> prioritizedTerritories,
-      final Map<Territory, Double> territoryValueMap) {
+      final List<Territory> clearedTerritories) {
 
     ProLogger.info("Check if we should try to hold attack territories");
 
     final ProOtherMoveOptions enemyAttackOptions = territoryManager.getEnemyAttackOptions();
     final Map<Territory, ProTerritory> attackMap = territoryManager.getAttackOptions().getTerritoryMap();
 
     // Determine which territories to try and hold
```

Comment:
I'm bit curious if it's easy to explain why the previous method was slow. I assume this added code is a re-write of the slow code.

Is this current/new code still somewhat slow? It looks like a candidate for parallelism. FWIW, this is equivalent down until `territoryValueMap`, and iterates in parallel

```
    // Determine which territories to try and hold
    final Set<Territory> territoriesToCheck = prioritizedTerritories.stream()
        .map(ProTerritory::getTerritory)
        .collect(Collectors.toSet());
    // add all nonAirAttackers from prioritized territories
    territoriesToCheck.addAll(
        prioritizedTerritories
            .parallelStream()
            .map(patd -> CollectionUtils.getMatches(patd.getMaxUnits(), Matches.unitIsNotAir()))
            .flatMap(Collection::stream)
            .map(nonAirAttacker -> ProData.unitTerritoryMap.get(nonAirAttacker))
            .collect(Collectors.toSet()));
```

Reply:
So this portion of the code is very fast and was added to determine which territories need to be passed to TerritoryValueUtils which is the slow method.